### PR TITLE
RoaringTreemap for a 64bit bitmap

### PR DESCRIPTION
**Pull request by bvinc**
Issue #38.  I just finished all of the operations I think the RoaringTreemap should have.  Here are my thoughts about all of the changes:

- Do you like the current module setup where it's in its own module and pulled in with `pub use` in lib.rs?
- Serialization isn't implemented.  I assume it's only defined for 32bit.
- The only thing messy and difficult was iteration.  It took me an hour to figure out that I needed `as _` to coerce a function item into a function pointer.
- Some of the ops really needed for me to be able to delete RoaringBitmaps from the tree as I'm iterating through the tree.  Obviously, rust doesn't like this, but less obviously, there is literally no way to do it with BTreeMap.  I decided to collect keys and make the deletion of entries a post-processing step at the end of the operation.  It sucks that the operation takes O(n) space when it doesn't have to, but compared to the actual bitmap size and the space savings after deleting, it's fine.
- The unit tests are kind of a mess.  Creating a subdirectory under /tests doesn't work.  I just duplicated the RoaringBitmap tests for now and they all pass.  They could probably use some more u64-oriented test data.

### CODE REVIEW

**Comment by Nemo157** [comment]
Awesome work 😄 

I've put a couple small issues I noticed going over the source.

> * Do you like the current module setup where it's in its own module and pulled in with pub use in lib.rs?

Yep, that looks like a great way to keep it separated. After merging this I may do the same for `RoaringBitmap` now that there's more than the one struct in this crate.

> * Serialization isn't implemented. I assume it's only defined for 32bit.

I didn't see any mention of it when I was implementing the 32bit version, so I'm fine with not supporting it till it's standardised.

> * The only thing messy and difficult was iteration. It took me an hour to figure out that I needed as _ to coerce a function item into a function pointer.

Uhhhh, yeah, I've spend way too long staring at that error message and not knowing what `rustc` wants me to do. I hope at some point that gets improved.

> * Some of the ops really needed for me to be able to delete RoaringBitmaps from the tree as I'm iterating through the tree. Obviously, rust doesn't like this, but less obviously, there is literally no way to do it with BTreeMap. I decided to collect keys and make the deletion of entries a post-processing step at the end of the operation. It sucks that the operation takes O(n) space when it doesn't have to, but compared to the actual bitmap size and the space savings after deleting, it's fine.

Yep, only other way I could see to implement it is to do something like `.into_iter().filter_map(|| do the operation).collect::<BTreeMap>()` which would almost certainly be less performant. The vector allocation should really not matter compared to how long the actual operation will take.

> * The unit tests are kind of a mess. Creating a subdirectory under /tests doesn't work. I just duplicated the RoaringBitmap tests for now and they all pass. They could probably use some more u64-oriented test data.

Huh, I guess I've never tried using subdirectories before; since that doesn't work having a `treemap_` prefix looks fine to me. I think having some tests that cover the removal of bitmaps during the ops would be good, I guess you can't test that directly, but you can at least set it up so the bitmap is removed directly instead of calling into it's operators and check the values aren't still there afterwards.

**Comment by bvinc** [comment]
> I've put a couple small issues I noticed going over the source.

If you meant to add some comments on the diff, I don't see them.

**Comment by Nemo157** [comment]
Ah, I need to "Request Changes" now, I think this is the first time I've used the new review system.

File: src/roaring_treemap/iter.rs [hunk 1]
```diff
@@ -0,0 +1,181 @@
...
+                                   fn((u32, RoaringBitmap)) -> To64IntoIter>;
+
+/// An iterator for `RoaringTreemap`.
+pub struct Iter<'a> {
+    inner: InnerIter<'a>,
+    size_hint: u64,
+}
+
+/// An iterator for `RoaringTreemap`.
+pub struct IntoIter {
```
Comment (bvinc) [inline src/roaring_treemap/iter.rs]: I think you're right.  I've decided to expose them at roaring_treemap::{Iter, IntoIter} by putting a `pub use` in roaring_treemap/mod.rs.  That way, it follows the model of std::collections::BTreeMap and std::collections::btree_map::Iter.  Does that sound good?  Maybe I should just keep RoaringTreemap in the module as well?

Don't you currently have the same problem with your IntoIter for RoaringBitmap?  I don't see any way it's exposed, but you're using it when you `impl IntoIterator for RoaringBitmap`.  I don't understand how that would work or how the documentation would work.

Ok, I just ran cargo doc, and yeah, it doesn't work.

**Comment by Nemo157** [comment]
Great, I'll make some time tomorrow to stick the existing code into a subdirectory as well, fix the iters not being exposed, then release a new version 🎉 (just realised I hadn't released a new version with my other changes yet, these are gonna be some hefty release notes).

**Pull request merged**